spontaneous reset support + sequencer fixes + SB-RMI support

bcantrill opened the pull request:
Fixes #1160 (among several other issues) -- and adds the necessary SP support oxidecomputer/stlouis#321.

Draft commit comment:

```
spontaneous reset support + sequencer fixes + SB-RMI support

- Adds logic to recognize a reset from host software -- along with a
  new state, A0Reset, that denotes that the host has reset.

- Fixes #1160 by having the RAA229618 driver always form its rail
  selection and subsequent read or write as a single transaction with
  respect to the I2C server

- Fixes #1135 by greatly increasing the busy spin count

- Fixes an issue whereby we were inadvertently backdriving via
  SP_TO_SP3_NMI_SYNC_FLOOD_L when in A2, preventing reset of the GPIO
  expander (manifesting itself as being in A0PlusHP on an
  A0PlusHP->A2->A0 transition rather than the expected A0).

- Cleans up the sequncer to always set CLT_RST in NIC_CTRL when in A2

- Adds an SB-RMI driver, interface and task to be able to query
  host CPU state 
```

mkeeter commented on drv/gimlet-seq-server/src/main.rs:
```diff
@@ -46,25 +47,42 @@ enum Trace {
 }
 
-ringbuf!(Trace, 64, Trace::None);
+ringbuf!(Trace, 100, Trace::None);
```
Nit: making this a power-of-two may be a little more efficient (mask versus division).

bcantrill replied:
It does -- but the deeper history has been really, really helpful for debugging sequencer issues.

bcantrill replied:
We had not had room for 128 originally, but with some of the other changes, now we do; bumping this to 128.

mkeeter commented on drv/i2c-devices/src/sbrmi.rs:
```diff
@@ -0,0 +1,339 @@
+        match err {
+            Error::BadRegisterRead { code, .. } => code,
+            Error::BadCpuidRead { code } => code,
+            _ => ResponseCode::BadResponse,
```
Doing a quick audit, it seems like different drivers do many different things when this conversion fails:
- `bmr491` (and many others) will panic
- `nvme_bmc` (and a few others) will return `ResponseCode::BadDeviceState`
- Here, we're returning `ResponseCode::BadResponse`

Philosophically, I think this should be `TryFrom`, but would require modifying I2C validation codegen.  I opened #1161 to track this.

bcantrill replied:
Thanks for filing that!

mkeeter commented on drv/i2c-devices/src/sbrmi.rs:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    fn read_reg(&self, reg: Register) -> Result<u8, Error> {
+        match self.device.read_reg::<u8, u8>(reg.into()) {
+            Ok(buf) => Ok(buf),
+            Err(code) => Err(Error::BadRegisterRead { reg, code }),
+        }
```
Nit:
```suggestion
        self.device
            .read_reg::<u8, u8>(reg.into())
            .map_err(|code| Error::BadRegisterRead { reg, code })
```

bcantrill replied:
Done

mkeeter commented on drv/i2c-devices/src/sbrmi.rs:
```diff
@@ -0,0 +1,339 @@
+
+    pub fn alert(&self) -> Result<[u8; 16], Error> {
+        //
+        // Have you ever played 52-bit Pickup?  Unlike the enabled bits, the
```
🤣

mkeeter commented on drv/i2c-devices/src/sbrmi.rs:
```diff
@@ -0,0 +1,339 @@
+        //   0x54      MceStat[3:0] = Threads[116,100,84,68]
+        //
+        let mut rval = [0u8; 16];
+        let banksize = 16usize;
+
+        for (bank, base) in [0x10, 0x50].iter().enumerate() {
```
Is it harmless to read extra registers that aren't in the table above (`0x55-0x5F`)?

If not, we could do something like
```suggestion
        for (bank, (base, banksize0)) in
            [(0x10, 16), (0x50, 5)].iter().enumerate()
        {
```

bcantrill replied:
Ha -- the comment was incomplete!  I had neglected to include the second page of the table; thanks for spotting that!

mkeeter commented on drv/sbrmi-api/src/lib.rs:
```diff
@@ -0,0 +1,59 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Client API for the MeanWell driver.
```
🤔

bcantrill replied:
🤦‍♂️

mkeeter commented on drv/sbrmi/src/main.rs:
```diff
@@ -0,0 +1,156 @@
+
+fn sbrmi() -> Sbrmi {
+    let devs = i2c_config::devices::sbrmi(I2C.get_task_id());
+    Sbrmi::new(&devs[0])
```
It would be a little more efficient to store this in the `ServerImpl`, so you don't need to rebuild it every time.

bcantrill replied:
Done

mkeeter commented:
I don't _love_ the timed-lease strategy ("timeouts, timeouts, always wrong...").

However, doing explicit lock + release is tricky: the release would naturally be part of the Idol API, but we want the `gimlet-seq` task to pause mid-transition, so we can't naturally return to the Idol dispatch.

Here's a quick sketch of an alternative:
- Replace `lease_devices` with `lock`, which stores the calling task ID in an `Option<TaskId>` within `gimlet-seq`
- Add a new `unlock` API, which sets that value to `None`
- During the critical period, if `gimlet-seq` is locked, then do a closed receive, _only_ listening for the `unlock` operation coming from the locking task (and faulting the caller if it tries to call any other IPCs, using `REPLY_FAULT`)
- Once we hear the `unlock` operation, then continue with the power-up sequence.  We know that other tasks won't be able to interfere with power-up, because they'll call `lock()`, and `gimlet-seq` won't return to Idol dispatch until it's done powering up.

If this seems plausible, [here's a `diff`](https://gist.github.com/mkeeter/ae085ae13a0c9207bb1e60cb37adb1a2) that implements this strategy.  It's totally untested, but I think it has the right ideas.

bcantrill commented:
> I don't _love_ the timed-lease strategy ("timeouts, timeouts, always wrong...").

While I agree in the abstract, we already have a bunch of time-based waiting in the sequencer path -- and I really want to avoid additional complexity here. So in this case, I'm not sure that the timeouts *are* wrong -- and in this case, I think they are quite a bit simpler than the alternative

cbiffle commented:
We had a brief call to talk through the requirements here, and I've suggested a different approach.

The core problem here, and the reason why the tasks are capable of achieving a destructive interleaving, is that operations on the Renesas VR require two I2C transactions:

1. A write to choose which rail to affect in a subsequent transaction.
2. Another write or write-read or read to affect that rail.

Our current I2C IPC API only exposes a single operation, write-then-read. However, we've hit requirements for more complex I2C transaction sequencing in the past, particularly around the Konami Code (which thankfully, afaict, we are not shipping). If, however, we were to extend the API to support a more general notion of uninterruptible transaction consisting of some combination of write and read phases, this would be incredibly general, and would fix the power-sequencer interleaving issue as a side effect -- as I understand it.

As a sketch of a totally-general operation to either replace or supplement write-then-read, consider:

- The new IPC message contains the target address, as it does today, but does not explicitly indicate write or read.
- The IPC can take an _arbitrary_ number of leases.
- The write bit on each lease determines whether it is a read (note: this appears backwards, because on an I2C read we must be able to write the lease, and vice versa).
- The operations described by the sequence of leases will be carried out one after the other without yielding the bus or switching muxes.

This would let you encode arbitrarily complex sequences of operations to the same I2C address as long as you don't require a STOP condition anywhere in the middle -- for instance, you could do "write 3 bytes, repeated start, write 2 bytes, repeated start, read 4 bytes, repeated start, write 1 byte, stop." The current write-then-read operation falls out of this as a special case.

Idol cannot currently express this operation, unfortunately, but the I2C server isn't implemented using Idol, so no big deal. (In fact, I was pretty sure we _used to have_ an operation like this, which is why the I2C server doesn't use Idol, but I may have been making that up.)

bcantrill commented on drv/i2c-devices/src/sbrmi.rs:
```diff
@@ -0,0 +1,339 @@
+        ecx: u32,
+    ) -> Result<CpuidResult, Error> {
+        let eax = eax.to_le_bytes();
+        let mut rval = CpuidResult {
+            ..Default::default()
+        };
```
Thank you!

bcantrill commented:
> We had a brief call to talk through the requirements here, and I've suggested a different approach.
> 
> The core problem here, and the reason why the tasks are capable of achieving a destructive interleaving, is that operations on the Renesas VR require two I2C transactions:
> 
> 1. A write to choose which rail to affect in a subsequent transaction.
> 2. Another write or write-read or read to affect that rail.
> 
> Our current I2C IPC API only exposes a single operation, write-then-read. However, we've hit requirements for more complex I2C transaction sequencing in the past, particularly around the Konami Code (which thankfully, afaict, we are not shipping). If, however, we were to extend the API to support a more general notion of uninterruptible transaction consisting of some combination of write and read phases, this would be incredibly general, and would fix the power-sequencer interleaving issue as a side effect -- as I understand it.
> 
> As a sketch of a totally-general operation to either replace or supplement write-then-read, consider:
> 
> * The new IPC message contains the target address, as it does today, but does not explicitly indicate write or read.
> * The IPC can take an _arbitrary_ number of leases.
> * The write bit on each lease determines whether it is a read (note: this appears backwards, because on an I2C read we must be able to write the lease, and vice versa).
> * The operations described by the sequence of leases will be carried out one after the other without yielding the bus or switching muxes.
> 
> This would let you encode arbitrarily complex sequences of operations to the same I2C address as long as you don't require a STOP condition anywhere in the middle -- for instance, you could do "write 3 bytes, repeated start, write 2 bytes, repeated start, read 4 bytes, repeated start, write 1 byte, stop." The current write-then-read operation falls out of this as a special case.
> 
> Idol cannot currently express this operation, unfortunately, but the I2C server isn't implemented using Idol, so no big deal. (In fact, I was pretty sure we _used to have_ an operation like this, which is why the I2C server doesn't use Idol, but I may have been making that up.)

I stopped short of making this truly arbitrary, but instead phrased it in terms of write/read pairs -- and allowing any number of these to be in a single I2C server transaction.  (This brought me straight into the grip of #1135, so I fixed that as well.)  I modified the RAA229618 and ISL68224 driver (the former being the source of the problem and the latter being a very similar driver), but stopped short of updating the mwocp68 driver:  we do not currently share this device across tasks (indeed, it's only used by the PSC) and it's quite a bit more involved to convert this.

 The draft commit comment has been updated accordingly!

mkeeter commented on drv/i2c-api/src/lib.rs:
```diff
@@ -510,6 +512,91 @@ impl I2cDevice {
+            response.as_bytes_mut(),
+            &[
+                Lease::from(buffer),
+                Lease::from(&empty[0..0]),
```
It would be nice to be able to write
```suggestion
                Lease::from(&[]),
```

To make that work, [insert this code right here](https://github.com/oxidecomputer/hubris/blob/master/sys/userlib/src/lib.rs#L94)
```rust
impl<'a> From<&'a [u8; 0]> for Lease<'a> {
    fn from(x: &'a [u8; 0]) -> Self {
        Self::read_only(x)
    }
}
```

mkeeter replied:
Alternatively,
```suggestion
                Lease::read_only(&[]),
```
works already, without need to add a new `impl From`

bcantrill replied:
I love that alternative -- will change this throughout that file.

mkeeter commented on drv/i2c-devices/src/isl68224.rs:
```diff
@@ -81,23 +81,16 @@ impl Isl68224 {
-        pmbus_write!(self.device, OPERATION, operation)
+        let mut op = pmbus_rail_read!(self.device, self.rail, OPERATION)?;
+        op.set_on_off_state(OPERATION::OnOffState::Off);
+        pmbus_write!(self.device, OPERATION, op)
```
Should this be `pmbus_rail_write!`?

bcantrill replied:
God, yes -- thank you!

mkeeter commented on drv/i2c-devices/src/isl68224.rs:
```diff
@@ -81,23 +81,16 @@ impl Isl68224 {
-        pmbus_write!(self.device, OPERATION, operation)
+        let mut op = pmbus_rail_read!(self.device, self.rail, OPERATION)?;
+        op.set_on_off_state(OPERATION::OnOffState::On);
+        pmbus_write!(self.device, OPERATION, op)
```
Same here

bcantrill replied:
Got it

mkeeter commented on drv/i2c-devices/src/lib.rs:
```diff
@@ -105,6 +150,40 @@ macro_rules! pmbus_write {
 
+macro_rules! pmbus_rail_write {
+    ($device:expr, $rail:expr, $dev:ident::$cmd:ident, $data:expr) => {{
+        let rpayload = [$dev::PAGE::CommandData::code(), $rail];
+
+        let mut payload = [0u8; $dev::$cmd::CommandData::len() + 1];
+        payload[0] = $dev::$cmd::CommandData::code();
+        $data.to_slice(&mut payload[1..]);
+
+        match $device.write_write(&rpayload, &payload) {
+            Err(code) => Err(Error::BadWrite {
+                cmd: $dev::$cmd::CommandData::code(),
+                code,
+            }),
+            Ok(_) => Ok(()),
+        }
```
I spent too much time trying to figure out way to remove this duplication:
```suggestion
        use $dev::{$cmd, PAGE};
        pmbus_rail_write!($device, $rail, $cmd, $data)
```

(the double `{{...}}` already make this is a local scope, so the `use` doesn't escape)

bcantrill replied:
That is a huge improvement!

mkeeter commented on drv/i2c-devices/src/lib.rs:
```diff
@@ -75,6 +76,50 @@ macro_rules! pmbus_read {
+    }};
+
+    ($device:expr, $rail:expr, $dev:ident::$cmd:ident) => {{
+        let payload = [$dev::PAGE::CommandData::code(), $rail];
+
+        match $dev::$cmd::CommandData::from_slice(&match $device
+            .write_read_reg::<u8, [u8; $dev::$cmd::CommandData::len()]>(
+                $dev::$cmd::CommandData::code(),
+                &payload,
+            ) {
+            Ok(rval) => Ok(rval),
+            Err(code) => Err(Error::BadRead {
+                cmd: $dev::$cmd::CommandData::code(),
+                code,
+            }),
+        }?) {
+            Some(data) => Ok(data),
+            None => Err(Error::BadData {
+                cmd: $dev::$cmd::CommandData::code(),
+            }),
+        }
```
```suggestion
        use $dev::{$cmd, PAGE};
        pmbus_rail_read!($device, $rail, $cmd)
```
(see explanation below; this could apply to every `pmbus_xyz!` macro)

bcantrill replied:
Done -- thank you again!
